challenge-manager/challenge-tree/generalized_path_timer.go

```diff
@@ -0,0 +1,264 @@
+type AncestorsQueryResponse struct {
+	AncestorLocalTimers []EdgeLocalTimer
+	AncestorEdgeIds     HonestAncestors
+}
+
+// ComputeHonestPathTimer for an honest edge at a block number given its ancestors'
+// local timers. It adds up all their values including the assertion
+// unrivaled timer and the edge's local timer.
+func (ht *HonestChallengeTree) ComputeHonestPathTimer(
+	ctx context.Context,
+	edgeId protocol.EdgeId,
+	ancestorLocalTimers []EdgeLocalTimer,
+	blockNumber uint64,
+) (PathTimer, error) {
+	edge, ok := ht.edges.TryGet(edgeId)
+	if !ok {
+		return 0, errNotFound(edgeId)
+	}
+	edgeLocalTimer, err := ht.localTimer(edge, blockNumber)
+	if err != nil {
+		return 0, err
+	}
+	total := PathTimer(edgeLocalTimer)
+	assertionUnrivaledTimer, err := ht.metadataReader.AssertionUnrivaledBlocks(
+		ctx, ht.topLevelAssertionHash,
+	)
+	if err != nil {
+		return 0, err
+	}
+	total += PathTimer(assertionUnrivaledTimer)
+	for _, timer := range ancestorLocalTimers {
+		total += PathTimer(timer)
+	}
+	return total, nil
+}
+
+// ComputeAncestorsWithTimers computes the ancestors of the given edge and their respective path timers, even
+// across challenge levels. Ancestor lists are linked through challenge levels via claimed edges. It is generalized
+// to any number of challenge levels in the protocol.
+func (ht *HonestChallengeTree) ComputeAncestorsWithTimers(
+	ctx context.Context,
+	edgeId protocol.EdgeId,
+	blockNumber uint64,
+) (*AncestorsQueryResponse, error) {
+	// First, checks if the edge exists before performing any computation.
+	startEdge, ok := ht.edges.TryGet(edgeId)
+	if !ok {
+		return nil, errNotFound(edgeId)
+	}
+	if ht.honestBlockChalLevelZeroEdge.IsNone() {
+		return nil, ErrNoHonestTopLevelEdge
+	}
+	currentChallengeLevel, err := startEdge.GetChallengeLevel()
+	if err != nil {
+		return nil, err
+	}
+
+	// Set a cursor at the edge we start from. We will update this cursor
+	// as we advance in this function.
+	currentEdge := startEdge
+
+	ancestry := make([]protocol.EdgeId, 0)
+	localTimers := make([]EdgeLocalTimer, 0)
+
+	// Challenge levels go from lowest to highest, where lowest is the smallest challenge level
+	// (where challenges are over individual, WASM opcodes). If we have 3 challenge levels,
+	// we will go from 0, 1, 2. We want the ancestors for an edge across an entire challenge
+	// tree, even across levels.
+	for currentChallengeLevel < protocol.ChallengeLevel(ht.totalChallengeLevels) {
+		// Compute the root edge for the current challenge level.
+		rootEdge, err := ht.honestRootAncestorAtChallengeLevel(currentEdge, currentChallengeLevel)
+		if err != nil {
+			return nil, err
+		}
+
+		// Compute the ancestors for the current edge in the current challenge level.
+		ancestorLocalTimers, ancestorsAtLevel, err := ht.findHonestAncestorsWithinChallengeLevel(ctx, rootEdge, currentEdge, blockNumber)
+		if err != nil {
+			return nil, err
+		}
+
+		// Advance the challenge level.
+		currentChallengeLevel += 1
```

Comment:
Move this increment next to `if currentChallengeLevel == protocol.ChallengeLevel(ht.totalChallengeLevels)`?